File: src/outliers.py
"""
outliers.py - Statistical outlier detection for TasteScope
Computes songs/artists that break the user's own rating patterns.
"""

from collections import defaultdict
from typing import List, Dict


def detect_outliers(rated_entries: list, all_artists: dict, ratings: list,
                    get_genre_dist_fn=None) -> Dict:
    """Compute all outlier categories and return structured data.

    Categories:
      artist_volatility  – artists with huge rating spread
      genre_rebels       – songs rated far from their genre average
      guilty_pleasures   – high-rated songs in usually-disliked genres
      disappointments    – low-rated songs in usually-loved genres
      one_hit_wonders    – one standout song far above an artist's other tracks
      rating_surprises   – songs 25+ pts away from your overall average
    """
    if not rated_entries or not ratings:
        return {'categories': {}, 'summary': {}}

    overall_avg = sum(ratings) / len(ratings) if ratings else 80

    # Build per-genre avg ratings
    genre_ratings = defaultdict(list)
    for r in rated_entries:
        g = r.get('_genre') or 'Uncategorized'
        if g != 'Uncategorized':
            genre_ratings[g].append(int(r['rating']))
    genre_avgs = {g: sum(rs)/len(rs) for g, rs in genre_ratings.items() if len(rs) >= 3}

    # ---- 1. Artist Volatility ----
    artist_volatility = []
    for artist, info in all_artists.items():
        rs = info.get('ratings', [])
        if len(rs) < 2:
            continue
        spread = max(rs) - min(rs)
        if spread >= 25:
            avg = sum(rs) / len(rs)
            artist_volatility.append({
                'artist': artist,
                'avg_rating': round(avg, 1),
                'min_rating': min(rs),
                'max_rating': max(rs),
                'spread': spread,
                'song_count': len(rs),
                'genre': info.get('genre', 'Uncategorized'),
            })
    artist_volatility.sort(key=lambda x: -x['spread'])

    # ---- 2. Genre Rebels (songs 20+ pts from genre avg) ----
    genre_rebels = []
    for r in rated_entries:
        g = r.get('_genre') or 'Uncategorized'
        if g not in genre_avgs:
            continue
        rating = int(r['rating'])
        diff = rating - genre_avgs[g]
        if abs(diff) >= 20:
            artists = _extract_artists_from_title(r.get('title', ''))
            genre_rebels.append({
                'title': r.get('title', ''),
                'artist': artists[0] if artists else '',
                'rating': rating,
                'genre': g,
                'genre_avg': round(genre_avgs[g], 1),
                'diff': round(diff, 1),
                'direction': 'above' if diff > 0 else 'below',
            })
    genre_rebels.sort(key=lambda x: -abs(x['diff']))

    # ---- 3. Guilty Pleasures (high-rated in low-rated genres) ----
    # Genres you avg <70 but rated a song 85+
    low_avg_genres = {g for g, avg in genre_avgs.items() if avg < 70}
    guilty_pleasures = []
    for r in rated_entries:
        g = r.get('_genre') or 'Uncategorized'
        if g not in low_avg_genres:
            continue
        rating = int(r['rating'])
        if rating >= 85:
            artists = _extract_artists_from_title(r.get('title', ''))
            guilty_pleasures.append({
                'title': r.get('title', ''),
                'artist': artists[0] if artists else '',
                'rating': rating,
                'genre': g,
                'genre_avg': round(genre_avgs.get(g, 0), 1),
            })
    guilty_pleasures.sort(key=lambda x: -x['rating'])

    # ---- 4. Disappointments (low-rated in high-rated genres) ----
    # Genres you avg >85 but rated a song <60
    high_avg_genres = {g for g, avg in genre_avgs.items() if avg > 85}
    disappointments = []
    for r in rated_entries:
        g = r.get('_genre') or 'Uncategorized'
        if g not in high_avg_genres:
            continue
        rating = int(r['rating'])
        if rating < 60:
            artists = _extract_artists_from_title(r.get('title', ''))
            disappointments.append({
                'title': r.get('title', ''),
                'artist': artists[0] if artists else '',
                'rating': rating,
                'genre': g,
                'genre_avg': round(genre_avgs.get(g, 0), 1),
            })
    disappointments.sort(key=lambda x: x['rating'])

    # ---- 5. One-Hit Wonders (one song 20+ pts above artist avg) ----
    one_hit_wonders = []
    for artist, info in all_artists.items():
        rs = info.get('ratings', [])
        if len(rs) < 2:
            continue
        other_avg = sum(rs) / len(rs)
        songs = info.get('songs', [])
        for song in songs:
            sr = int(song.get('rating', 0))
            if sr - other_avg >= 20:
                one_hit_wonders.append({
                    'title': song.get('title', ''),
                    'artist': artist,
                    'rating': sr,
                    'artist_avg': round(other_avg, 1),
                    'diff': round(sr - other_avg, 1),
                    'genre': info.get('genre', 'Uncategorized'),
                })
    one_hit_wonders.sort(key=lambda x: -x['diff'])

    # ---- 6. Rating Surprises (25+ pts from your overall avg) ----
    rating_surprises = []
    for r in rated_entries:
        rating = int(r['rating'])
        diff = rating - overall_avg
        if abs(diff) >= 25:
            artists = _extract_artists_from_title(r.get('title', ''))
            genre = r.get('_genre') or 'Uncategorized'
            rating_surprises.append({
                'title': r.get('title', ''),
                'artist': artists[0] if artists else '',
                'rating': rating,
                'overall_avg': round(overall_avg, 1),
                'diff': round(diff, 1),
                'direction': 'above' if diff > 0 else 'below',
                'genre': genre,
            })
    rating_surprises.sort(key=lambda x: -abs(x['diff']))

    return {
        'categories': {
            'artist_volatility': artist_volatility[:15],
            'genre_rebels': genre_rebels[:20],
            'guilty_pleasures': guilty_pleasures[:15],
            'disappointments': disappointments[:15],
            'one_hit_wonders': one_hit_wonders[:15],
            'rating_surprises': rating_surprises[:20],
        },
        'summary': {
            'overall_avg': round(overall_avg, 1),
            'total_songs': len(rated_entries),
            'total_artists': len([a for a in all_artists if all_artists[a].get('ratings')]),
            'volatile_artists': len(artist_volatility),
            'genre_rebels_count': len(genre_rebels),
            'guilty_pleasures_count': len(guilty_pleasures),
            'disappointments_count': len(disappointments),
            'one_hit_wonders_count': len(one_hit_wonders),
            'surprises_count': len(rating_surprises),
        },
    }


def _extract_artists_from_title(title: str) -> list:
    """Extract artist names from 'Song (Artist, Year)' or 'Artist - Song' patterns."""
    import re
    if not title:
        return []
    # Pattern: "Song Name (Artist, Year)"
    m = re.search(r'\(([^,]+),\s*\d{4}\)', title)
    if m:
        return [m.group(1).strip()]
    # Pattern: "Artist - Song" or "Artist – Song"
    m = re.match(r'^(.+?)\s*[-–—]\s*.+', title)
    if m:
        return [m.group(1).strip()]
    return []

File: static/js/outliers.js
/**
 * outliers.js - Statistical Outliers View
 * Shows songs/artists that break your own rating patterns.
 */

let _outliersData = null;

async function loadOutliers() {
    try {
        const res = await fetch('/api/outliers');
        if (!res.ok) throw new Error('Failed to load outliers');
        _outliersData = await res.json();
        renderOutliers(_outliersData);
    } catch (err) {
        console.error('Outliers load error:', err);
        document.getElementById('outliersGrid').innerHTML =
            '<div class="loading-msg" style="color:var(--danger)">Error loading outliers. Is the server running?</div>';
    }
}

function renderOutliers(data) {
    const { categories: cats, summary } = data;
    const summaryEl = document.getElementById('outliersSummary');
    const gridEl = document.getElementById('outliersGrid');

    // Summary bar
    summaryEl.innerHTML = `
        <div class="outliers-stats-row">
            <div class="outlier-stat"><span class="outlier-stat-val">${summary.volatile_artists}</span><span class="outlier-stat-lbl">Volatile Artists</span></div>
            <div class="outlier-stat"><span class="outlier-stat-val">${summary.genre_rebels_count}</span><span class="outlier-stat-lbl">Genre Rebels</span></div>
            <div class="outlier-stat"><span class="outlier-stat-val">${summary.guilty_pleasures_count}</span><span class="outlier-stat-lbl">Guilty Pleasures</span></div>
            <div class="outlier-stat"><span class="outlier-stat-val">${summary.disappointments_count}</span><span class="outlier-stat-lbl">Disappointments</span></div>
            <div class="outlier-stat"><span class="outlier-stat-val">${summary.one_hit_wonders_count}</span><span class="outlier-stat-lbl">One-Hit Wonders</span></div>
            <div class="outlier-stat"><span class="outlier-stat-val">${summary.surprises_count}</span><span class="outlier-stat-lbl">Rating Surprises</span></div>
        </div>
    `;

    // Build cards
    let html = '';

    // 1. Artist Volatility
    if (cats.artist_volatility && cats.artist_volatility.length > 0) {
        html += outlierSection('🌋 Artist Volatility', 'Artists where your ratings swing wildly',
            cats.artist_volatility.map(a => `
                <div class="outlier-card">
                    <div class="outlier-card-header">
                        <span class="outlier-name">${esc(a.artist)}</span>
                        <span class="outlier-badge">${a.genre}</span>
                    </div>
                    <div class="outlier-bar-container">
                        <div class="outlier-bar-range">
                            <div class="outlier-bar-fill" style="left:${a.min_rating}%;width:${a.max_rating - a.min_rating}%"></div>
                            <div class="outlier-bar-avg" style="left:${a.avg_rating}%"></div>
                        </div>
                        <div class="outlier-bar-labels">
                            <span>${a.min_rating}</span>
                            <span class="outlier-avg-label">avg ${a.avg_rating}</span>
                            <span>${a.max_rating}</span>
                        </div>
                    </div>
                    <div class="outlier-detail">${a.song_count} songs · spread of <strong>${a.spread} pts</strong></div>
                </div>
            `).join('')
        );
    }

    // 2. Guilty Pleasures
    if (cats.guilty_pleasures && cats.guilty_pleasures.length > 0) {
        html += outlierSection('😏 Guilty Pleasures', 'High-rated songs in genres you usually score low',
            cats.guilty_pleasures.map(s => songCard(s, `You avg ${s.genre_avg} in ${s.genre} but rated this ${s.rating}`)).join('')
        );
    }

    // 3. Disappointments
    if (cats.disappointments && cats.disappointments.length > 0) {
        html += outlierSection('😬 Disappointments', 'Low-rated songs in genres you usually love',
            cats.disappointments.map(s => songCard(s, `You avg ${s.genre_avg} in ${s.genre} but rated this only ${s.rating}`)).join('')
        );
    }

    // 4. One-Hit Wonders
    if (cats.one_hit_wonders && cats.one_hit_wonders.length > 0) {
        html += outlierSection('⭐ One-Hit Wonders', 'One standout song far above an artist\'s average',
            cats.one_hit_wonders.map(s => `
                <div class="outlier-card">
                    <div class="outlier-card-header">
                        <span class="outlier-name">${esc(s.title)}</span>
                        <span class="outlier-badge outlier-badge-up">+${s.diff} above avg</span>
                    </div>
                    <div class="outlier-detail">${esc(s.artist)} · ${s.genre}</div>
                    <div class="outlier-rating-row">
                        <span class="outlier-rating">${s.rating}/100</span>
                        <span class="outlier-vs">vs artist avg ${s.artist_avg}</span>
                    </div>
                </div>
            `).join('')
        );
    }

    // 5. Genre Rebels
    if (cats.genre_rebels && cats.genre_rebels.length > 0) {
        html += outlierSection('🔥 Genre Rebels', 'Songs rated 20+ pts away from their genre average',
            cats.genre_rebels.slice(0, 12).map(s => `
                <div class="outlier-card">
                    <div class="outlier-card-header">
                        <span class="outlier-name">${esc(s.title)}</span>
                        <span class="outlier-badge ${s.direction === 'above' ? 'outlier-badge-up' : 'outlier-badge-down'}">${s.diff > 0 ? '+' : ''}${s.diff} vs ${s.genre}</span>
                    </div>
                    <div class="outlier-detail">${esc(s.artist)} · ${s.genre}</div>
                    <div class="outlier-rating-row">
                        <span class="outlier-rating">${s.rating}/100</span>
                        <span class="outlier-vs">genre avg ${s.genre_avg}</span>
                    </div>
                </div>
            `).join('')
        );
    }

    // 6. Rating Surprises
    if (cats.rating_surprises && cats.rating_surprises.length > 0) {
        const above = cats.rating_surprises.filter(s => s.direction === 'above').slice(0, 8);
        const below = cats.rating_surprises.filter(s => s.direction === 'below').slice(0, 8);
        let surpriseHtml = '';
        if (above.length > 0) {
            surpriseHtml += '<h4 class="outlier-subhead">📈 Far Above Your Average</h4>';
            surpriseHtml += above.map(s => `
                <div class="outlier-card outlier-card-up">
                    <div class="outlier-card-header">
                        <span class="outlier-name">${esc(s.title)}</span>
                        <span class="outlier-badge outlier-badge-up">+${s.diff}</span>
                    </div>
                    <div class="outlier-detail">${esc(s.artist)} · ${s.genre}</div>
                    <div class="outlier-rating-row">
                        <span class="outlier-rating">${s.rating}/100</span>
                        <span class="outlier-vs">your avg ${s.overall_avg}</span>
                    </div>
                </div>
            `).join('');
        }
        if (below.length > 0) {
            surpriseHtml += '<h4 class="outlier-subhead">📉 Far Below Your Average</h4>';
            surpriseHtml += below.map(s => `
                <div class="outlier-card outlier-card-down">
                    <div class="outlier-card-header">
                        <span class="outlier-name">${esc(s.title)}</span>
                        <span class="outlier-badge outlier-badge-down">${s.diff}</span>
                    </div>
                    <div class="outlier-detail">${esc(s.artist)} · ${s.genre}</div>
                    <div class="outlier-rating-row">
                        <span class="outlier-rating">${s.rating}/100</span>
                        <span class="outlier-vs">your avg ${s.overall_avg}</span>
                    </div>
                </div>
            `).join('');
        }
        html += outlierSection('🎲 Rating Surprises', 'Songs 25+ pts from your overall average', surpriseHtml);
    }

    if (!html) {
        html = '<div class="loading-msg">Not enough data to detect outliers yet. Rate more songs!</div>';
    }

    gridEl.innerHTML = html;
}

// ---- Helpers ----

function outlierSection(title, subtitle, content) {
    return `
        <div class="outlier-section">
            <div class="outlier-section-header">
                <h3>${title}</h3>
                <p class="outlier-section-sub">${subtitle}</p>
            </div>
            <div class="outlier-cards">${content}</div>
        </div>
    `;
}

function songCard(s, reason) {
    return `
        <div class="outlier-card">
            <div class="outlier-card-header">
                <span class="outlier-name">${esc(s.title)}</span>
                <span class="outlier-badge">${s.genre}</span>
            </div>
            <div class="outlier-detail">${esc(s.artist)} · ${reason}</div>
            <div class="outlier-rating-row">
                <span class="outlier-rating">${s.rating}/100</span>
                <span class="outlier-vs">genre avg ${s.genre_avg}</span>
            </div>
        </div>
    `;
}

function esc(str) {
    if (!str) return '';
    const el = document.createElement('span');
    el.textContent = str;
    return el.innerHTML;
}
